File: Problems/python/boj_친구.py
#https://www.acmicpc.net/problem/1058
#Floyd-Warshall

import sys
input = sys.stdin.readline
inf = int(1e9)

n = int(input())
mat = [list(input()) for _ in range(n)]

graph = [[0]*n for _ in range(n)]
for k in range(n):
    for i in range(n):
        for j in range(n):
            if i==j: continue
            if mat[i][j]=='Y' or (mat[i][k]=='Y' and mat[k][j]=='Y'):
                graph[i][j] = 1

result = 0
for i in graph:
    count = i.count(1)
    result = max(result, count)

print(result)


'''
문제를 보고 플로이드 워셜이라는것은 알았다.
하지만 문제에서 요구하는 값을 카운팅 하는것이 어려웠다.
-> 대칭행렬이니깐 각 행을 사람이라 치면 2-친구의 수는 그 행의 1의 개수와 같다.
'''

File: Problems/cpp/boj_친구.cpp
//https://www.acmicpc.net/problem/1058
//Floyd-Warshall

#include <algorithm>
#include <iostream>
#include <string>
#include <memory.h>

using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n;
    cin >> n;

    char mat[n][n];
    int graph[n][n];
    string temp;
    for (int i=0; i<n; i++){
        cin >> temp;
        for (int j=0; j<n; j++){
            if (temp[j]=='Y')
                mat[i][j] = 'Y';
            else
                mat[i][j] = 'N';
        }
    }
    
    memset(graph, 0, sizeof(graph)); //배열 0으로 초기화 꼭 해줘야함
    
    //여기서 중요한건 2다리까지만 구하는 거다!!!
    for (int k=0; k<n; k++){
        for (int i=0; i<n; i++){
            for (int j=0; j<n; j++){
                if (i==j) continue;
                if (mat[i][j]=='Y' || (mat[i][k]=='Y' && mat[k][j]=='Y')){
                    graph[i][j] = 1;
                }
            }
        }
    }

    int result = 0;
    for (int i=0; i<n; i++){
        int cnt = count(graph[i], graph[i]+n, 1);
        result = max(result, cnt);
    }
    
    cout << result;
}

/*
2다리까지만 구해야 하는 문제였는데 다 구하니깐 안됐었다.=> (i,i) 빼고 다 1됨.
floyd-warshall이 2다리까지만 구하는 것!
*/